packages/unified-selection/src/unified-selection/HiliteSetProvider.ts

```diff
@@ -178,53 +162,41 @@ export class HiliteSetProvider {
+                      WHERE r.TargetECClassId IS (BisCore.SubCategory) AND ${this.formBindings("s.ECInstanceId", categoryKeys, bindings)}
                     ),
                     SubCategories (ECInstanceId) AS (
                       SELECT s.ECInstanceId as \`ECInstanceId\` FROM BisCore.SubCategory s
-                      WHERE InVirtualSet(?, s.ECInstanceId)
+                      WHERE ${this.formBindings("s.ECInstanceId", subCategoryKeys, bindings)}
                     )
-                   SELECT DISTINCT ECInstanceId FROM CategorySubCategories
+                   SELECT ECInstanceId FROM CategorySubCategories
                    UNION
-                   SELECT DISTINCT ECInstanceId FROM SubCategories;`;
+                   SELECT ECInstanceId FROM SubCategories;`;
 
     return this.executeQuery(query, bindings);
   }
 
   private async getHilitedElements(map: Map<InstanceIdType, string[]>): Promise<string[]> {
     const groupInformationElementKeys = map.get("groupInformationElement") ?? [];
-    let geometricElementKeySet = new Set(map.get("geometricElement") ?? []);
+    const geometricElementKeys = map.get("geometricElement") ?? [];
+    const functionalElements = map.get("functionalElement") ?? [];
     const elementKeys = map.get("element") ?? [];
+    const hasFunctionalElements = functionalElements.length !== 0;
 
-    if (this._hasFunctionalSchema) {
-      const functionalGeometricElements = new Set(await this.getHilitedFunctionalElements(map));
-      geometricElementKeySet = new Set([...geometricElementKeySet, ...functionalGeometricElements]);
-    }
-
-    const geometricElementKeys = Array.from(geometricElementKeySet);
-
-    if (groupInformationElementKeys.length === 0 && geometricElementKeys.length === 0 && elementKeys.length === 0) {
+    if (groupInformationElementKeys.length === 0 && geometricElementKeys.length === 0 && elementKeys.length === 0 && !hasFunctionalElements) {
       return [];
     }
 
-    const bindings: ECSqlBinding[] = [
-      { type: "idset", value: groupInformationElementKeys },
-      { type: "idset", value: elementKeys },
-      { type: "idset", value: geometricElementKeys },
-    ];
+    const bindings: ECSqlBinding[] = [];
     const query = `WITH RECURSIVE
+                    ${hasFunctionalElements ? this.getHilitedFunctionalElementsQuery(functionalElements, bindings) : ""}
                     GroupMembers (ECInstanceId, ECClassId) AS (
-                      SELECT r.TargetECInstanceId, r.TargetECClassId FROM BisCore.GroupInformationElement s
-                        JOIN BisCore.ElementGroupsMembers r ON r.SourceECInstanceId = s.ECInstanceId
-                      WHERE InVirtualSet(?, s.ECInstanceId)
+                      SELECT TargetECInstanceId, TargetECClassId FROM BisCore.ElementGroupsMembers
+                      WHERE ${this.formBindings("ECInstanceId", groupInformationElementKeys, bindings)}
```

Comment:
I believe we should be using `SourceECInstanceId` here. If I'm right, it suggests we're missing a full stack tests for this case.